Sources/Core/Shared/Operation.swift

```diff
@@ -569,33 +684,69 @@ public extension Operation {
      - parameter errors: an array of `ErrorType`, which defaults to empty.
      */
     final func finish(receivedErrors: [ErrorType] = []) {
-        if !_hasFinishedAlready {
-            _hasFinishedAlready = true
+        _finish(receivedErrors, fromCancel: false)
+    }
+
+    private final func _finish(receivedErrors: [ErrorType], fromCancel: Bool = false) {
+        let willFinish = stateLock.withCriticalScope { _ -> Bool in
+            // Do not finish if already finished or finishing
+            guard state <= .Finishing else { return false }
+            // Only a single call to _finish should continue
+            // (.cancel() sets _isHandlingFinish and fromCancel=true, if appropriate.)
+            guard !_isHandlingFinish || fromCancel else { return false }
+            _isHandlingFinish = true
+            return true
+        }
+
+        guard willFinish else { return }
+
+        // NOTE:
+        // - The stateLock should only be held when necessary, and should not
+        //   be held when notifying observers (whether via KVO or Operation's
+        //   observers) or deadlock can result.
+
+        let changedExecutingState = executing
+        if changedExecutingState {
+            willChangeValueForKey(NSOperationKeyPaths.Executing.rawValue)
+        }
+
+        stateLock.withCriticalScope {
             state = .Finishing
+        }
 
+        if changedExecutingState {
+            didChangeValueForKey(NSOperationKeyPaths.Executing.rawValue)
+        }
+
+        let errors = stateLock.withCriticalScope { () -> [ErrorType] in
             _internalErrors.appendContentsOf(receivedErrors)
-            operationDidFinish(_internalErrors)
+            return _internalErrors
+        }
 
-            if errors.isEmpty {
-                log.verbose("Will finish with no errors.")
-            }
-            else {
-                log.warning("Will finish with \(errors.count) errors.")
-            }
+        if errors.isEmpty {
+            log.verbose("Will finish with no errors.")
+        }
+        else {
+            log.warning("Will finish with \(errors.count) errors.")
+        }
 
-            willFinishObservers.forEach { $0.willFinishOperation(self, errors: self._internalErrors) }
+        operationWillFinish(errors)
+        willChangeValueForKey(NSOperationKeyPaths.Finished.rawValue)
+        willFinishObservers.forEach { $0.willFinishOperation(self, errors: errors) }
 
-            state = .Finished
+        state = .Finished
```

Comment:
Not needed because of a side-effect of the setter of Operation.`state` (which acquires the lock), but I agree that it's much better to be explicit about it (and not doing so is an oversight). Fixing.